dd-sdk-android-core/src/main/kotlin/com/datadog/android/core/internal/persistence/datastore/DatastoreFileReader.kt

```diff
@@ -0,0 +1,130 @@
+            return
+        }
+
+        readFromDataStoreFile(datastoreFile, deserializer, tlvBlockFileReader, version, callback)
+    }
+
+    @Suppress("ReturnCount", "ThreadSafety")
+    private fun <T : Any> readFromDataStoreFile(
+        datastoreFile: File,
+        deserializer: Deserializer<String, T>,
+        tlvBlockFileReader: TLVBlockFileReader,
+        requestedVersion: Int?,
+        callback: DataStoreCallback<T>
+    ) {
+        val tlvBlocks = tlvBlockFileReader.read(datastoreFile)
+
+        // there should be as many blocks read as there are block types
+        val numberBlocksFound = tlvBlocks.size
+        val numberBlocksExpected = TLVBlockType.values().size
+        if (numberBlocksFound != numberBlocksExpected) {
+            logInvalidNumberOfBlocksError(numberBlocksFound, numberBlocksExpected)
+            callback.onFailure()
+            return
+        }
+
+        val dataStoreContent = mapToDataStoreContents(deserializer, tlvBlocks)
+
+        if (dataStoreContent == null) {
+            callback.onFailure()
+            return
+        }
+
+        // if an optional version is specified then only return data if the entry version exactly matches
+        if (requestedVersion != null && requestedVersion != dataStoreContent.versionCode) {
+            callback.onSuccess(null)
+            return
+        }
+
+        callback.onSuccess(dataStoreContent)
+    }
+
+    private fun <T : Any> mapToDataStoreContents(
+        deserializer: Deserializer<String, T>,
+        tlvBlocks: List<TLVBlock>
+    ): DataStoreContent<T>? {
+        if (tlvBlocks[0].type != TLVBlockType.VERSION_CODE &&
+            tlvBlocks[1].type != TLVBlockType.DATA
+        ) {
+            logBlocksInUnexpectedBlocksOrderError()
+            return null
+        }
+
+        val versionCodeBlock = tlvBlocks[0]
+        val dataBlock = tlvBlocks[1]
+
+        return DataStoreContent(
+            versionCode = versionCodeBlock.data.toInt(),
+            data = deserializer.deserialize(String(dataBlock.data))
+        )
+    }
+
+    private fun logInvalidNumberOfBlocksError(numberBlocksFound: Int, numberBlocksExpected: Int) {
+        internalLogger.log(
+            level = InternalLogger.Level.ERROR,
+            target = InternalLogger.Target.MAINTAINER,
+            messageBuilder = {
+                INVALID_NUMBER_OF_BLOCKS_ERROR
+                    .format(Locale.US, numberBlocksFound, numberBlocksExpected)
+            }
+        )
+    }
+
+    private fun logBlocksInUnexpectedBlocksOrderError() {
+        internalLogger.log(
+            level = InternalLogger.Level.ERROR,
+            target = InternalLogger.Target.MAINTAINER,
+            messageBuilder = { UNEXPECTED_BLOCKS_ORDER_ERROR }
+        )
+    }
+
+    internal companion object {
+        internal const val INVALID_NUMBER_OF_BLOCKS_ERROR =
+            "Read error - datastore entry has invalid number of blocks. Was: %s expected: %s"
```

Comment:
```suggestion
            "Read error - datastore entry has invalid number of blocks. Was: %s, expected: %s"
```

Also shouldn't it be `%d` here instead of `%s`, given that arguments are numbers? But probably for integers it doesn't matter.